File: src/main/java/com/baulsupp/oksocial/services/twitter/TwurlCredentialsStore.java
package com.baulsupp.oksocial.services.twitter;

import com.baulsupp.oksocial.credentials.CredentialsStore;
import com.baulsupp.oksocial.credentials.ServiceDefinition;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.PropertyNamingStrategy;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
import com.google.common.base.Throwables;
import java.io.File;
import java.util.Optional;

public class TwurlCredentialsStore implements CredentialsStore<TwitterCredentials> {
  private final File file;

  private ServiceDefinition<TwitterCredentials> serviceDefinition = new TwitterServiceDefinition();

  public TwurlCredentialsStore(File file) {
    this.file = file;
  }

  public TwurlRc readTwurlRc() {
    try {
      if (file.isFile()) {
        ObjectMapper objectMapper = new ObjectMapper(new YAMLFactory());
        objectMapper.setPropertyNamingStrategy(
            PropertyNamingStrategy.CAMEL_CASE_TO_LOWER_CASE_WITH_UNDERSCORES);

        return objectMapper.readValue(file, TwurlRc.class);
      } else {
        return null;
      }
    } catch (Exception e) {
      throw Throwables.propagate(e);
    }
  }

  @Override
  public ServiceDefinition<TwitterCredentials> getServiceDefinition() {
    return serviceDefinition;
  }

  public Optional<TwitterCredentials> readDefaultCredentials() {
    TwurlRc twurlRc = readTwurlRc();

    if (twurlRc != null) {
      String username = twurlRc.defaultProfile().get(0);
      String consumerKey = twurlRc.defaultProfile().get(1);

      return Optional.ofNullable(twurlRc.readCredentials(username, consumerKey));
    } else {
      return Optional.empty();
    }
  }

  @Override
  public void storeCredentials(TwitterCredentials credentials) {
    throw new UnsupportedOperationException();
  }
}

File: src/main/kotlin/com/baulsupp/okurl/services/twitter/twurlrc/TwurlCredentialsStore.kt
package com.baulsupp.okurl.services.twitter.twurlrc

import com.baulsupp.okurl.services.twitter.TwitterCredentials
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.PropertyNamingStrategy
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
import java.io.File

class TwurlCredentialsStore(val file: File) {

  fun readTwurlRc(): TwurlRc? = try {
    if (file.isFile) {
      val objectMapper = ObjectMapper(YAMLFactory())
      objectMapper.propertyNamingStrategy = PropertyNamingStrategy.SNAKE_CASE

      objectMapper.readValue(file, TwurlRc::class.java)
    } else {
      null
    }
  } catch (e: Exception) {
    throw RuntimeException(e)
  }

  fun readCredentials(): TwitterCredentials? = readTwurlRc()?.let {
    val username = it.defaultProfile()[0]
    val consumerKey = it.defaultProfile()[1]

    it.readCredentials(username, consumerKey)
  }

  companion object {
    var TWURL_STORE = TwurlCredentialsStore(File(System.getProperty("user.home"), ".twurlrc"))
  }
}
